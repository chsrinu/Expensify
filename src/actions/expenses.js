import uuid from 'uuid';

export const AddExpense = ({
    description = 'testDescription',
    note = 'testNote',
    amount = 1,
    createdAt = 0
    } = {}
) => (
        {
            type: 'EXPENSE',
            expense : {
                id: uuid(),
                description: description,
                note: note,
                amount: amount,
                createdAt: createdAt
            }
        }
    )
export const editExpense = (id,updates) => (
    {
        type:'EDIT_EXPENSE',
        id:id,
        updates:updates
    }
)
export const removeExpense = ({id}) => {
    console.log(id);
    return {
        type: 'REMOVE_EXPENSE',
        id: id
    }
}