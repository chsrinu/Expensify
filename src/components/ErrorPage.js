import React from 'react';

const ErrorPage = () => (
    <div>
        <h1>404</h1>
        <h2>You bugger there's no route here </h2>
    </div>
)

export default ErrorPage;