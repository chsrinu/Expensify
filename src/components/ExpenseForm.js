import React from 'react'
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state={
            description: props.expense ? props.expense.description:'',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount:'',
            createdAt: props.expense ? moment(props.expense.createdAt):moment(),
            calendarFocus:false
        }
    }
    onDateChange = (createdAt) => {
        //console.log('date has changed');
        if(createdAt)
            this.setState({createdAt:createdAt})
    }
    descriptionChangeListner = (e) => {
        let desc=e.target.value;
        //console.log(desc);
        this.setState({description:desc})
    }
    noteChangeListner = (e) => {
        let note = e.target.value;
        this.setState({note:note})
    }
    amountChangeListner = (e) => {
        let amount = e.target.value;
        //console.log(amount);
        if(!amount || amount.match(/^\d*(\.\d{0,2})?$/))
            this.setState({amount:amount});
    }
    onSubmit = (e) => {
        e.preventDefault();
        console.log('clicked the button');
        this.props.onSubmit(
            {...this.props.expense,
                description:this.state.description,
                note:this.state.note,
                amount:this.state.amount,
                createdAt:this.state.createdAt.valueOf()
            })
    } 
    render(){
    return (
        <div>
        <form onSubmit = {this.onSubmit}>
            <input 
                type='text'
                value = {this.state.description}
                placeholder='Description'
                onChange = {this.descriptionChangeListner}
                autoFocus
                />
            <input 
                type='text'
                placeholder='Amount'
                value={this.state.amount} 
                onChange = {this.amountChangeListner}
                />
            <SingleDatePicker
                date={this.state.createdAt} // momentPropTypes.momentObj or null
                onDateChange={this.onDateChange} // PropTypes.func.isRequired
                focused={this.state.calendarFocus} // PropTypes.bool
                onFocusChange={({ focused }) => this.setState({calendarFocus:focused })} // PropTypes.func.isRequired
                numberOfMonths={1}
                isOutsideRange={()=>false}
                />
            <textarea
                placeholder = 'Add note for your expense(optional)'
                onChange = {this.noteChangeListner}    
            />
            <button>Add Expense</button>
        </form>
    </div>
    )
    }
}
export default ExpenseForm