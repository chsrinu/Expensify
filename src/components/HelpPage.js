import React from 'react';

const HelpPage = () => (
    <h3>I am here to help you </h3>
);

export default HelpPage;