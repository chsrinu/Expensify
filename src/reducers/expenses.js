export const expensesReducer = (state=[],action) => {
    switch(action.type){
        case 'EXPENSE':
            return [...state,action.expense];
        case 'REMOVE_EXPENSE':
           // console.log('deleting expense',action.id);
            return (state.filter((item)=>{
                if(item.id!=action.id)
                    return item;
            }))
        case 'EDIT_EXPENSE':
            console.log(action.id, action.updates);
            return state.map((expense)=>{
                if(expense.id===action.id)
                    return {...expense,...action.updates}
                return expense;
            })
        default:
        return state;
    }
}