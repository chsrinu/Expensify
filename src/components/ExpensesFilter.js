import React from 'react';
import {connect } from 'react-redux';
import {setTextFilter,sortByDate,sortByAmount, setStartDate,setEndDate} from '../actions/filters';
import { DateRangePicker } from 'react-dates'
//import {setStartDate, setEndDate} from '../actions/filters';
import moment from 'moment';

export class ExpensesFilter extends React.Component {
state = {focused:null}
dateChangeListner = ({startDate, endDate}) => {
    console.log('dates are changed',startDate);
        //this.props.dispatch(setStartDate(startDate));
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
        //this.props.dispatch(setEndDate(endDate));
}
render(){
return(
    <div>
        <input type='text' value={this.props.filter.text} onChange={(e)=>{
            console.log('filter value is ',e.target.value);
        //this.props.dispatch(setTextFilter(e.target.value));
        this.props.setTextFilter(e.target.value);
    }}/>
        <select id='selectFilter' value = { this.props.filter.sortBy } onChange = {(e)=>{
            let filter = e.target.value 
            console.log(filter);
            switch(filter){
                case 'amount':
                   // this.props.dispatch(sortByAmount());
                   this.props.sortByAmount();
                    break;
                case 'date':
                   // this.props.dispatch(sortByDate());
                   this.props.sortByDate();
                    break;
            }
        }}>
            <option value='amount'>Amount</option>
            <option value='date'>Date</option>
        </select>
        <DateRangePicker
            startDate={this.props.filter.startDate} // momentPropTypes.momentObj or null,
            startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
            endDate={this.props.filter.endDate} // momentPropTypes.momentObj or null,
            endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
            onDatesChange={(startDate, endDate) => this.dateChangeListner(startDate,endDate)} // PropTypes.func.isRequired,
            focusedInput= {this.state.focused}
            onFocusChange={(focused) => this.setState({focused })} // PropTypes.func.isRequired    
            numberOfMonths={1}
            showClearDates
            isOutsideRange={()=>false}
            />
    </div>
)
}
}
const mapStateToProps = (state) => (
    {filter:state.filters}
)
const mapDispatchToProps = (dispatch) => (
    {
        setStartDate : ({startDate}) => dispatch(setStartDate(startDate)),
        setEndDate : ({endDate}) => dispatch(setEndDate(endDate)),
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByAmount: () => dispatch(sortByAmount),
        sortByDate: () => dispatch(sortByDate)
    }
)
export default connect(mapStateToProps)(ExpensesFilter);