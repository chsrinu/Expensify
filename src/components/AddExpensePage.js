import React from 'react';
import ExpenseForm from './ExpenseForm';
import {AddExpense} from '../actions/expenses';
import { connect } from 'react-redux';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.onSubmit(expense);
        this.props.history.push('/');
    }
    render(){
        return (
            <div>
                <h2>Add Expense </h2>
            <ExpenseForm 
                onSubmit = {this.onSubmit} />
            </div> 
        )
    }
}
const mapDispatchToProps = (dispatch) => (
    {
        onSubmit: (expense) => dispatch(AddExpense(expense))
    }
)

export default connect(undefined, mapDispatchToProps)(AddExpensePage);