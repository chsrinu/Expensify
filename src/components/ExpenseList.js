import React from 'react';
import {connect} from 'react-redux'; 
import ExpenseListItem from './ExpenseListItem'
import ExpensesFilter from './ExpensesFilter'
import {visibilitySettings} from '../selectors/expenses.js'

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        <ExpensesFilter />
        {
            props.list.length > 0 ?
            (props.list.map((item)=><ExpenseListItem item={item} key={item.id} />))
            :
            (<p>No expenses To show</p>)
        }
    </div>
)
const mapStateToProps = (state) => (
    {
        list:visibilitySettings(state.expenses,state.filters)
    }
)
export default connect(mapStateToProps)(ExpenseList)