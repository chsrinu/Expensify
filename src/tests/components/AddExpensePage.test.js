import React from 'react';
import {shallow } from 'enzyme';
import {AddExpensePage} from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses'

test('AddExpensePage testCase',() => {
    const onSubmit = jest.fn();
    const history = { push: jest.fn()}
    const wrapper = shallow(<AddExpensePage onSubmit = {onSubmit} history = {history} />)
    expect(wrapper).toMatchSnapshot();
})


test('AddExpensePage with onSubmit spy testCase',() => {
    const onSubmit = jest.fn();
    const history = { push: jest.fn()}
    const wrapper = shallow(<AddExpensePage onSubmit = {onSubmit} history = {history} />)
    
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(onSubmit).toHaveBeenLastCalledWith(expenses[1]);
})