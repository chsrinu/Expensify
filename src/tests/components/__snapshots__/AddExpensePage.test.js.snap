// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddExpensePage testCase 1`] = `
<div>
  <h2>
    Add Expense 
  </h2>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
